doc/step-in-groovy.md

```diff
@@ -0,0 +1,116 @@
+# Implementing Step in Groovy
+The standard way of implementing a step is by writing Java code that extends from `Step` and `StepDescriptor`,
+but you can also define a step in Jenkins Pipeline Script.
+
+This approach is quite convenient when your step composes other
+existing steps. For example, the following files define
+the equivalent of the `retry` step that repeats the body up to several times:
+
+```java
+// src/main/java/acme/RetryStep.java
+package acme;
+
+class RetryStep extends GroovyStep {
+    private int times;
+
+    @DataBoundConstructor
+    public RetryStep(int times) {
+        this.times = times;
+    }
+
+    public int getTimes() {
+        return times;
+    }
+
+    @Extension
+    public static class DescriptorImpl extends GroovyStepDescriptor {
+        @Override public String getFunctionName() {
+	        return "myRetryClone";
+        }
+
+        @Override public String getDisplayName() {
+	        return "Retry step";
+        }
+
+        @Override
+        public boolean takesImplicitBlockArgument() {
+            return true;
+        }
+    }
+}
+```
+
+```groovy
+// src/main/resources/acme/RetryStepExecution.groovy
+package acme;
+
+import org.jenkinsci.plugins.workflow.cps.steps.ingroovy.GroovyStepExecution
+
+public class RetryStepExecution extends GroovyStepExecution {
+    public void call(Closure body) {
+        int attempt = 0;
+        while (true) {
+            try {
+                echo "trying ${attempt}"
+                body();
+                return;
+            } catch (Exception e) {
+                if (attempt++ < step.times) {
+                    continue; // retry
+                } else {
+                    throw e;
+                }
+            }
+        }
+    }
+}
+```
+
+From Jenkins Pipeline, you can use this like you'd use any other steps:
+
+```
+myRetryClone(3) {
+  sh './flakyTest.sh'
+}
+```
+
+Note the directory in which the files are defined. `GroovyStep` and `GroovyStepDescriptor`
+are defined in `src/main/java` and is written just like any other steps.
+`GroovyStepExecution` is defined in `src/main/resources` to be packaged into the plugin
+in the source form, so that it can run in the CPS groovy interpreter at the runtime.
+
+The `call` method on `GroovyStepExecution` subtype defines the actual logic of step.
```

Comment:
I just asked whether it's possible to return a value from a `GroovyStepExecution` along the lines of, say, `ReadFileExecution` - it'd be nice to have some explanation of what the return type of `call` can be.